### PR TITLE
Allow readonly array as parameter of .choices()

### PR DESCRIPTION
**Pull request by ouuan**
## Problem

Without `readonly`, readonly arrays cannot be passed as parameters.

Example: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMAjACgIYCcsC5EDOUWMYA5gNoC6AlIgN4C+AUKJLAsgEyY75YBTDABMEAGwCehYqUq0GLZhARFE2LCkQBeRBQDkGPVQDcSlVDU4u23QaNqCiZWCKnW6dShqngPT9-deLC4A3yCQ4yA

## Solution

Add `readonly` in the TypeScript typings.

## ChangeLog

Allow readonly arrays as parameters of `.choices()`.

### CODE REVIEW

**Comment by shadowspawn** [comment]
I have been doing a bit of reading and still thinking about this. Just changing the type does allow passing a ReadonlyArray as parameter, and reflects the intent that the call itself will not change the array, and lets the caller prevent casual changes to their reference to the array. However, the Commander code effectively removes the readonly in the implementation, so it is weak!

The array is stored in `argChoices`:

```
argChoices?: string[];
```

```js
choices(values) {
  this.argChoices = values;
  ...
  }
```

The readonly could be honoured by taking a defensive copy of the parameter:

```js
choices(values) {
  this.argChoices = Array.isArray(values) ? values.slice() : values;
  ...
  }
```

Reference:
- https://www.typescriptlang.org/docs/handbook/2/objects.html#the-readonlyarray-type

Related:
- `aliases()`, but in this case the implementation does take a copy

**Comment by shadowspawn** [comment]
I don't think readonly can be used for parameters in JSDoc?

- https://jsdoc.app/tags-type.html
- https://jsdoc.app/tags-readonly.html

**Comment by ouuan** [comment]
> ```js
> this.argChoices = Array.isArray(values) ? values.slice() : values;
> ```

Could you elaborate on why `Array.isArray(values)` can be false?

**Comment by shadowspawn** [comment]
> Could you elaborate on why Array.isArray(values) can be false?

Hmm, good question. Probably not needed! I was being paranoid, and had in mind perhaps passing `undefined` or `null` to remove the choices, although that is not supported either in the code or in the typings.

**Comment by shadowspawn** [comment]
For interest, what is your use case? (Why do you have a readonly array of choices?)

**Comment by ouuan** [comment]
> For interest, what is your use case? (Why do you have a readonly array of choices?)

I use a library that provides a union type for valid options. The library doesn't provide an array for the valid options, so I created a readonly array in my own code that contains these options and type-checked the array against the union type provided by the library.

File: lib/argument.js [hunk 1]
```diff
@@ -97,7 +97,7 @@ class Argument {
    */
 
   choices(values) {
-    this.argChoices = values;
+    this.argChoices = values.slice();
     this.parseArg = (arg, previous) => {
       if (!values.includes(arg)) {
```
Comment (shadowspawn) [inline lib/argument.js:102]: The uses of `values` in the parse function should be replaced with `this.argChoices` now they may be different.

File: lib/option.js [hunk 2]
```diff
@@ -117,7 +117,7 @@ class Option {
    */
 
   choices(values) {
-    this.argChoices = values;
+    this.argChoices = values.slice();
     this.parseArg = (arg, previous) => {
       if (!values.includes(arg)) {
```
Comment (shadowspawn) [inline lib/option.js:122]: The uses of `values` in the parse function should be replaced with `this.argChoices` now they may be different.

**Comment by shadowspawn** [comment]
I offered some tests in a PR back to the branch to check the parameter is actually treated as readonly at runtime.

**Review by shadowspawn — APPROVED** [review]
I don't think this will be widely needed and requires taking a defensive copy of param, but do I like making the TypeScript API more usable (readonly) and safer (defensive copy).

Thumbs up from me. 👍

**Review by abetomo — APPROVED** [review]
👍

**Comment by shadowspawn** [comment]
Commander v9 has been released.

**Review by Sammiegirl411 — COMMENTED** [review]
Let's do this

**Pull request merged**